src/clock.zig

```diff
@@ -0,0 +1,235 @@
+    }
+
+    fn synchronize_window(self: *Clock) ?Marzullo.Interval {
+        const window_elapsed = monotonic_timestamp() - self.window_epoch_monotonic;
+        if (window_elapsed < self.window_duration_min) return null;
+
+        // Starting with the most clock offset tolerance, while we have a majority, find the best
+        // smallest interval with the least clock offset tolerance, reducing tolerance at each step:
+        var best: ?Marzullo.Interval = null;
+
+        var tolerance = self.clock_offset_max;
+        while (tolerance > 0) : (tolerance /= 16) {
+            var interval = Marzullo.smallest_interval(self.window_tuples(tolerance));
+            const majority = interval.sources_true > @divTrunc(self.sources.len, 2);
+            if (!majority) break;
+            best = interval;
+        }
+
+        return best;
+    }
+
+    fn window_tuples(self: *Clock, clock_offset_max: u64) []Marzullo.Tuple {
+        assert(self.sources[self.replica].?.clock_offset == 0);
+        assert(self.sources[self.replica].?.one_way_delay == 0);
+        var count: usize = 0;
+        for (self.sources) |sampled, source| {
+            if (sampled) |sample| {
+                const range = @intCast(i64, sample.one_way_delay) + @intCast(i64, clock_offset_max);
+                self.marzullo_tuples[count] = Marzullo.Tuple{
+                    .source = @intCast(u8, source),
+                    .offset = sample.clock_offset - range,
+                    .bound = .lower,
+                };
+                count += 1;
+                self.marzullo_tuples[count] = Marzullo.Tuple{
+                    .source = @intCast(u8, source),
+                    .offset = sample.clock_offset + range,
+                    .bound = .upper,
+                };
+                count += 1;
+            }
+        }
+        return self.marzullo_tuples[0..count];
+    }
+};
+
+// TODO Each Clock instance should accept a Time instance that can be ticked deterministically for
+// testing purposes, and which will expose the following two timestamps:
+// This is the same pattern we use for Journal (all the logic) and Storage (the raw read/write).
+
+/// A timestamp to measure elapsed time, meaningful only on the same system, not across reboots.
+/// Always use a monotonic timestamp if the goal is to measure elapsed time.
+/// This clock is not affected by discontinuous jumps in the system time, for example if the
+/// system administrator manually changes the clock.
+fn monotonic_timestamp() u64 {
+    // The true monotonic clock on Linux is not in fact CLOCK_MONOTONIC:
+    // CLOCK_MONOTONIC excludes elapsed time while the system is suspended (e.g. VM migration).
+    // CLOCK_BOOTTIME is the same as CLOCK_MONOTONIC but includes elapsed time during a suspend.
+    // For more detail and why CLOCK_MONOTONIC_RAW is even worse than CLOCK_MONOTONIC,
+    // see https://github.com/ziglang/zig/pull/933#discussion_r656021295.
```

Comment:
The situation is even worse, as there are known to be hardware and/or software bugs that make the monotonic clock not actually monotonic on several platforms as discovered by Rust:
https://github.com/rust-lang/rust/blob/ac3e680193a3e11e60b07840ffb1db12793de110/library/std/src/time.rs#L223-L250

Not sure if this affects us as we don't support the listed hardware yet, but it's something we should be aware of.

Reply:
Fantastic! "Never trust the kernel"

Enjoyed the humor in that comment there:

> To hopefully mitigate the impact of this, a few platforms are excluded as "these at least haven't gone backwards yet".

I was planning to add an assertion that the monotonic clock is actually monotonic.

Reply:
Here's another interesting example: https://bugzilla.redhat.com/show_bug.cgi?id=448449